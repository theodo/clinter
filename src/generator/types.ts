import { Linter } from "eslint";
import { ProjectInfoObject } from "types";

export type ESLintGenerator = (userAnswers: ProjectInfoObject) => (config: Linter.Config) => Linter.Config;

export type ESLintDependencyGenerator = (userAnswers: ProjectInfoObject) => (deps: string[]) => string[];

export interface ESLintOverrider {
  files: string[];
}
