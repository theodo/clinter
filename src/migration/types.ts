export interface MigrationResults {
  fixableErrorCount: number;
  errorCount: number;
}
