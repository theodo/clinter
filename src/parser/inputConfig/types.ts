import { AnswerObject, ModeAnswerObject } from "types";

export interface InputConfig {
  generatorConfig: AnswerObject;
  modeConfig: ModeAnswerObject;
}
