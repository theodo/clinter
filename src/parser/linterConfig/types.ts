import { Options } from "prettier";
import { Linter } from "eslint";

export enum FileExtension {
  JSON = "json",
  YAML = "yml",
  JS = "js",
  NONE = "none",
}

export interface ConfigFileObject {
  name: string;
  extension: FileExtension;
  attribute?: string;
}

export type LinterConfigs = Options | Linter.Config;

export interface ConfigContainer<T extends LinterConfigs> {
  config: T;
  file: ConfigFileObject;
}
