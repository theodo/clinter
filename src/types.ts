export enum FormatterInfo {
  None = "None",
  Prettier = "Prettier",
}

export interface FormatterInfoObject {
  formatter: FormatterInfo;
}

export enum FrontFrameworkInfo {
  None = "None",
  Vue = "Vue",
  React = "React",
}

export interface FrontFrameworkInfoObject {
  frontFramework: FrontFrameworkInfo;
}

export enum TypescriptInfo {
  None = "None",
  WithTypeChecking = "With type checking",
  NoTypeChecking = "Without type checking",
}

export interface TypescriptInfoObject {
  typescript: TypescriptInfo;
}

export enum EnvInfo {
  Browser = "Browser",
  Node = "Node",
}

export interface EnvInfoObject {
  env: EnvInfo[];
}

export enum TestFrameworkInfo {
  None = "None",
  Jest = "Jest",
}

export interface TestFrameworkInfoObject {
  test: TestFrameworkInfo;
}

export type ProjectInfoObject = FormatterInfoObject &
  FrontFrameworkInfoObject &
  TypescriptInfoObject &
  EnvInfoObject &
  TestFrameworkInfoObject;

export enum ClinterModeInfo {
  Generator = "Generator Mode",
  Upgrade = "Upgrade Mode",
}

export interface ClinterModeInfoObject {
  mode: ClinterModeInfo;
}

export interface MigrationModeInfoObject {
  migration: boolean;
}

export interface DependenciesConfigObject {
  upgradeDependencies: boolean;
}

export interface ClinterSettings {
  generatorConfig: ProjectInfoObject;
  modeConfig: ClinterModeInfoObject;
  migrationModeConfig: MigrationModeInfoObject;
  dependenciesConfig: DependenciesConfigObject;
}
