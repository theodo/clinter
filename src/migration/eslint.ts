import { ESLint, Linter } from "eslint";
import fs from "fs";
import { MigrationResults } from "migration/types";

const getErrors = async (dirPath: string): Promise<ESLint.LintResult[]> => {
  const eslint = new ESLint({
    cwd: dirPath,
  });
  const results = await eslint.lintFiles("**/*.{js,ts,jsx,tsx}");

  return results;
};

const groupErrorsByLine = (errors: Linter.LintMessage[]): Record<number, Linter.LintMessage[]> => {
  return errors.reduce((result: Record<number, Linter.LintMessage[]>, error) => {
    const lineNumber = error.line;
    if (result[lineNumber] === undefined) {
      result[lineNumber] = [];
    }

    result[lineNumber].push(error);
    return result;
  }, {});
};

const insertErrorIgnoreLines = (sourceLines: string[], errorsByLine: Record<number, Linter.LintMessage[]>): string[] =>
  ((Object.entries(errorsByLine) as unknown) as [number, Linter.LintMessage[]][])
    .sort(([lineA], [lineB]) => lineB - lineA)
    .reduce((source, [lineNumber, errors]) => {
      const mappedErrors = errors
        .filter((error) => error.severity === 2 && error.fix === undefined)
        .map((error) => error.ruleId);

      if (mappedErrors.length === 0) {
        return source;
      }

      const errorRulesString = mappedErrors.join(", ");

      const ignoreString = `// eslint-disable-next-line ${errorRulesString}`;

      const indentation = /^\s*/.exec(source[lineNumber - 1])?.[0] ?? "";

      source.splice(lineNumber - 1, 0, indentation + ignoreString);

      return source;
    }, sourceLines);

const insertIgnoreLinesInFile = (filePath: string, fileErrors: Linter.LintMessage[]) => {
  const fileSource = fs.readFileSync(filePath).toString().split("\n");

  const errorsByLine = groupErrorsByLine(fileErrors);

  const newSource = insertErrorIgnoreLines(fileSource, errorsByLine).join("\n");

  fs.writeFileSync(filePath, newSource);
};

export const migrateProject = async (dirPath: string): Promise<MigrationResults> => {
  const errors = await getErrors(dirPath);

  return errors.reduce(
    (results: MigrationResults, error) => {
      insertIgnoreLinesInFile(error.filePath, error.messages);

      return {
        errorCount: results.errorCount + error.errorCount,
        fixableErrorCount: results.fixableErrorCount + error.fixableErrorCount,
      };
    },
    {
      errorCount: 0,
      fixableErrorCount: 0,
    }
  );
};
