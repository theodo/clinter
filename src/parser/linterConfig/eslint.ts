import { Linter } from "eslint";
import { ConfigContainer, ConfigFileObject, FileExtension } from "parser/linterConfig/types";
import { parseLinterConfigFiles } from "parser/linterConfig/parseLinterConfigFiles";

const possibleESLintFiles: ConfigFileObject[] = [
  { name: ".eslintrc", extension: FileExtension.NONE },
  { name: ".eslintrc.js", extension: FileExtension.JS },
  { name: ".eslintrc.json", extension: FileExtension.JSON },
  { name: ".eslintrc.yaml", extension: FileExtension.YAML },
  { name: "package.json", extension: FileExtension.JSON, attribute: "eslintConfig" },
];

export function findESLintConfigurationFiles(dirPath: string): ConfigContainer<Linter.Config>[] {
  return parseLinterConfigFiles(possibleESLintFiles, dirPath);
}
