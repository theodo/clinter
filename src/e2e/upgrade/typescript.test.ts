import { Linter } from "eslint";

import { defaultInputConfig, makeTestService } from "e2e/helpers";
import { ClinterModeInfo, TypescriptInfo } from "types";

describe("ESLint Base Configuration Upgrade Mode", () => {
  it("should return a typescript eslint configuration with an override and no typescript-eslint in base rules", async () => {
    const testService = makeTestService();

    const initialConfig: Linter.Config = {
      extends: [
        "plugin:@typescript-eslint/recommended",
        "plugin:@typescript-eslint/recommended-requiring-type-checking",
      ],
      rules: {
        "@typescript-eslint/prefer-optional-chain": "error",
        "@typescript-eslint/prefer-nullish-coalescing": "error",
        "@typescript-eslint/strict-boolean-expressions": "error",
      },
    };

    testService.loadInputConfig({
      ...defaultInputConfig,
      modeConfig: { mode: ClinterModeInfo.Upgrade },
      generatorConfig: { ...defaultInputConfig.generatorConfig, typescript: TypescriptInfo.WithTypeChecking },
    });
    testService.loadInitialLinterConfig(JSON.stringify(initialConfig));
    testService.loadInitialLinterConfig;
    await testService.runClinter();
    const outputConfig = testService.getOutputConfig();

    expect(outputConfig).toMatchSnapshot(`eslint-typescript-upgrade`);
  });

  it("should return a typescript eslint configuration with no type checking with an override and no typescript-eslint in base rules", async () => {
    const testService = makeTestService();

    const initialConfig: Linter.Config = {
      extends: [
        "plugin:@typescript-eslint/recommended",
        "plugin:@typescript-eslint/recommended-requiring-type-checking",
      ],
      rules: {
        "@typescript-eslint/prefer-optional-chain": "error",
      },
    };

    testService.loadInputConfig({
      ...defaultInputConfig,
      modeConfig: { mode: ClinterModeInfo.Upgrade },
      generatorConfig: { ...defaultInputConfig.generatorConfig, typescript: TypescriptInfo.NoTypeChecking },
    });
    testService.loadInitialLinterConfig(JSON.stringify(initialConfig));
    testService.loadInitialLinterConfig;
    await testService.runClinter();
    const outputConfig = testService.getOutputConfig();

    expect(outputConfig).toMatchSnapshot(`eslint-typescript-upgrade-no-type-checking`);
  });
});
