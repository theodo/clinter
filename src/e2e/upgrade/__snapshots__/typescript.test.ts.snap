// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ESLint Base Configuration Upgrade Mode should return a typescript eslint configuration with an override and no typescript-eslint in base rules: eslint-typescript-upgrade 1`] = `
"{
  \\"extends\\": [
    \\"eslint:recommended\\"
  ],
  \\"rules\\": {
    \\"curly\\": [
      \\"error\\",
      \\"all\\"
    ],
    \\"eqeqeq\\": [
      \\"error\\",
      \\"smart\\"
    ],
    \\"complexity\\": [
      \\"error\\",
      8
    ],
    \\"import/no-extraneous-dependencies\\": [
      \\"error\\",
      {
        \\"devDependencies\\": true,
        \\"optionalDependencies\\": false,
        \\"peerDependencies\\": false
      }
    ],
    \\"no-shadow\\": [
      \\"error\\",
      {
        \\"hoist\\": \\"all\\"
      }
    ],
    \\"prefer-const\\": \\"error\\",
    \\"import/order\\": [
      \\"error\\",
      {
        \\"groups\\": [
          [
            \\"external\\",
            \\"builtin\\"
          ],
          \\"internal\\",
          [
            \\"parent\\",
            \\"sibling\\",
            \\"index\\"
          ]
        ]
      }
    ],
    \\"sort-imports\\": [
      \\"error\\",
      {
        \\"ignoreCase\\": true,
        \\"ignoreDeclarationSort\\": true,
        \\"ignoreMemberSort\\": false,
        \\"memberSyntaxSortOrder\\": [
          \\"none\\",
          \\"all\\",
          \\"multiple\\",
          \\"single\\"
        ]
      }
    ],
    \\"padding-line-between-statements\\": [
      \\"error\\",
      {
        \\"blankLine\\": \\"always\\",
        \\"prev\\": \\"*\\",
        \\"next\\": \\"return\\"
      }
    ]
  },
  \\"root\\": true,
  \\"plugins\\": [
    \\"import\\"
  ],
  \\"overrides\\": [
    {
      \\"files\\": [
        \\"**/*.ts?(x)\\"
      ],
      \\"extends\\": [
        \\"plugin:@typescript-eslint/recommended\\",
        \\"plugin:@typescript-eslint/recommended-requiring-type-checking\\"
      ],
      \\"parser\\": \\"@typescript-eslint/parser\\",
      \\"parserOptions\\": {
        \\"project\\": \\"tsconfig.json\\"
      },
      \\"rules\\": {
        \\"@typescript-eslint/prefer-optional-chain\\": \\"error\\",
        \\"no-shadow\\": \\"off\\",
        \\"@typescript-eslint/no-shadow\\": \\"error\\",
        \\"@typescript-eslint/prefer-nullish-coalescing\\": \\"error\\",
        \\"@typescript-eslint/strict-boolean-expressions\\": \\"error\\"
      }
    }
  ]
}"
`;

exports[`ESLint Base Configuration Upgrade Mode should return a typescript eslint configuration with no type checking with an override and no typescript-eslint in base rules: eslint-typescript-upgrade-no-type-checking 1`] = `
"{
  \\"extends\\": [
    \\"eslint:recommended\\"
  ],
  \\"rules\\": {
    \\"curly\\": [
      \\"error\\",
      \\"all\\"
    ],
    \\"eqeqeq\\": [
      \\"error\\",
      \\"smart\\"
    ],
    \\"complexity\\": [
      \\"error\\",
      8
    ],
    \\"import/no-extraneous-dependencies\\": [
      \\"error\\",
      {
        \\"devDependencies\\": true,
        \\"optionalDependencies\\": false,
        \\"peerDependencies\\": false
      }
    ],
    \\"no-shadow\\": [
      \\"error\\",
      {
        \\"hoist\\": \\"all\\"
      }
    ],
    \\"prefer-const\\": \\"error\\",
    \\"import/order\\": [
      \\"error\\",
      {
        \\"groups\\": [
          [
            \\"external\\",
            \\"builtin\\"
          ],
          \\"internal\\",
          [
            \\"parent\\",
            \\"sibling\\",
            \\"index\\"
          ]
        ]
      }
    ],
    \\"sort-imports\\": [
      \\"error\\",
      {
        \\"ignoreCase\\": true,
        \\"ignoreDeclarationSort\\": true,
        \\"ignoreMemberSort\\": false,
        \\"memberSyntaxSortOrder\\": [
          \\"none\\",
          \\"all\\",
          \\"multiple\\",
          \\"single\\"
        ]
      }
    ],
    \\"padding-line-between-statements\\": [
      \\"error\\",
      {
        \\"blankLine\\": \\"always\\",
        \\"prev\\": \\"*\\",
        \\"next\\": \\"return\\"
      }
    ]
  },
  \\"root\\": true,
  \\"plugins\\": [
    \\"import\\"
  ],
  \\"overrides\\": [
    {
      \\"files\\": [
        \\"**/*.ts?(x)\\"
      ],
      \\"extends\\": [
        \\"plugin:@typescript-eslint/recommended\\"
      ],
      \\"parser\\": \\"@typescript-eslint/parser\\",
      \\"parserOptions\\": {
        \\"project\\": \\"tsconfig.json\\"
      },
      \\"rules\\": {
        \\"@typescript-eslint/prefer-optional-chain\\": \\"error\\",
        \\"no-shadow\\": \\"off\\",
        \\"@typescript-eslint/no-shadow\\": \\"error\\"
      }
    }
  ]
}"
`;
